machine/util/callbacks/early_stopping_callback.py

```diff
@@ -0,0 +1,94 @@
+from machine.util.callbacks import Callback
+
+
+class EarlyStoppingCallback(Callback):
+    """
+    Original callback taken from https://github.com/ncullen93/torchsample
+    Early Stopping to terminate training early under certain conditions
+    """
+
+    def __init__(self,
+                 monitor='eval_losses',
+                 lm_name=None,
+                 min_delta=0,
+                 patience=5,
+                 minimize=True):
+        """
+        EarlyStopping callback to exit the training loop if training or
+        validation loss or metric does not improve by a certain amount
+        for a certain number of epochs
+        Args:
+            monitor : string in {'eval_losses', 'eval_metrics', 'train_losses', 'train_metrics'}
+                whether to monitor train or val loss
+            lm_name: loss or metric name eg. 'Avg NLLoss' or 'Word Accuracy'
+                    If not specified then the first element
+                    in the monitor array is used
+                    (default None)
+            min_delta : float
+                minimum change in monitored value to qualify as improvement.
+                This number should be positive.
+            patience : integer
+                number of epochs to wait for improvment before terminating.
+                the counter be reset after each improvment
+            minimize: minimize quantity, if false then early stopping will maximize
+        """
+        if 'loss' in monitor:
+            self.loss = True
+        elif 'metric' in monitor:
+            self.loss = False
+        else:
+            raise ValueError(
+                "Monitor must be string in {'eval_losses', \
+                'eval_metrics', 'train_losses', 'train_metrics'}")
+
+        self.monitor = monitor
+        self.lm_name = lm_name
+        self.minimize = minimize
+
+        self.min_delta = min_delta
+        self.patience = patience
+        self.wait = 0
+        self.best_lm = 1e-15
+        super(EarlyStoppingCallback, self).__init__()
+
+    def on_train_begin(self, info=None):
+        self.wait = 0
+        if self.minimize:
+            self.best_lm = 1e15
+        else:
+            self.best_lm = -1e15
+
+    def on_epoch_end(self, info=None):
+        """
+        Function called at the end of every epoch
+        This allows specifing what eval or train loss to use
+        """
+
+        # If specific loss/metric name is specified
+        if self.lm_name is not None:
```

Comment:
This block has to fetch the correct value for the early stopping, right? Is a `break` really necessary to do that? To make the code more easily understandable, perhaps this block can be be moved into a separate function `_get_loss_values` or something like that? Does that make sense?

Reply:
Break is there so we don't check every metric or loss but stop as soon as we have a match. I think this is pretty clear, but I can split it to add a check as well.

Reply:
I understood what it meant, but I thought it would maybe be cleaner to add it to a separate function. Not very important though!